File: tests/reboot/std/item/v1/item_tests.ts
// `eslint` disabled to preserve separation of imports for documentation.
// eslint-disable-next-line
import { Any, createRegistry } from "@bufbuild/protobuf";
// eslint-disable-next-line
import { Value } from "@bufbuild/protobuf";
// eslint-disable-next-line
import { JsonObject } from "@bufbuild/protobuf";
import { strict as assert } from "node:assert";
import { test } from "node:test";
import { CreateRequest as SomeMessage } from "../../../greeter_rbt.js";

test("test item", async (t) => {
  await t.test("Item with value", async (t) => {
    // Test storing a value in an `Item`.
    // `console.log` statements are used as part of the documentation examples.
    let item;
    item = { value: Value.fromJson(true) };
    console.log(item.value?.toJson());

    item = { value: Value.fromJson(1) };
    console.log(item.value?.toJson());

    item = { value: Value.fromJson("apple") };
    console.log(item.value?.toJson());

    item = { value: Value.fromJson(["a", "b", "c"]) };
    console.log(item.value?.toJson());

    item = { value: Value.fromJson({ details: "details-go-here" }) };
    console.log(item.value?.toJson());

    let json = item.value?.toJson() as JsonObject;
    assert(json["details"] === "details-go-here");
  });

  await t.test("Item with bytes", async (t) => {
    // Test storing bytes in an `Item`.
    // `console.log` statements are used as part of the documentation examples.
    const item = { bytes: new TextEncoder().encode("some-bytes") };
    console.log(item.bytes);

    assert(new TextDecoder().decode(item.bytes) === "some-bytes");
  });

  await t.test("Item with any", async (t) => {
    // Test storing bytes in an `Item`.
    // `console.log` statements are used as part of the documentation examples.
    const contents = {
      title: "king",
      name: "nemo",
      adjective: "fishy",
    };
    const message = new SomeMessage(contents);

    const item = { any: Any.pack(message) };

    const registry = createRegistry(SomeMessage);
    const unpackedMessage = item.any.unpack(registry) as SomeMessage;

    assert(unpackedMessage.name === "nemo");
  });
});

File: tests/reboot/std/item/v1/item_tests.py
import unittest

# Import used in Item documentation.
# isort: off
from reboot.protobuf import (
    as_bool,
    as_dict,
    as_int,
    as_list,
    as_str,
    from_bool,
    from_dict,
    from_int,
    from_list,
    from_str,
)
from reboot.protobuf import pack, unpack
# isort: on
from reboot.std.item.v1.item import Item
from tests.reboot.greeter_rbt import CreateRequest as SomeMessage


class TestCase(unittest.IsolatedAsyncioTestCase):

    async def test_item_with_value(self) -> None:
        """
        Test storing a value in an `Item`. `print`s are for documentation examples.
        """
        item = Item(value=from_bool(True))
        print(as_bool(item.value))

        item = Item(value=from_int(3))
        print(as_int(item.value))

        item = Item(value=from_str("apple"))
        print(as_str(item.value))

        item = Item(value=from_list(["a", "b", "c"]))
        print(as_list(item.value))

        item = Item(value=from_dict({"details": "details-go-here"}))
        print(as_dict(item.value))

        self.assertEquals(as_dict(item.value)["details"], "details-go-here")

    async def test_item_with_bytes(self) -> None:
        """
        Test storing bytes in an `Item`. `print`s are for documentation examples.
        """
        item = Item(bytes=b"some-bytes")
        print(item.bytes)

        self.assertEqual(b"some-bytes", item.bytes)

    async def test_item_with_any(self) -> None:
        """
        Test storing an `Any` in an `Item`. `print`s are for documentation examples.
        """
        contents = {
            "title": "king",
            "name": "nemo",
            "adjective": "fishy",
        }

        message = SomeMessage(**contents)

        item = Item(any=pack(message))

        unpacked_message = unpack(item.any, SomeMessage)

        self.assertEqual(unpacked_message.title, "king")


if __name__ == '__main__':
    unittest.main()
